### PR TITLE
Refactor Key Format

### PR DESCRIPTION
**Pull request by sideninja**
Closes: #232

The key refactor changes the format of the keys as described in the issue.

It provides better flexibility and consistency.

### CODE REVIEW

File: internal/emulator/start.go [hunk 1]
```diff
@@ -77,23 +77,22 @@ func ConfiguredServiceKey(
...
+	serviceAccount, err := proj.EmulatorServiceAccount()
+	if err != nil {
+		util.Exit(1, err.Error())
+	}
 
-	serviceKeyHex, ok := serviceAccount.DefaultKey().(*project.HexAccountKey)
-	if !ok {
+	privateKey, err := serviceAccount.Key().PrivateKey()
+	if err != nil {
 		util.Exit(1, "Only hexadecimal keys can be used as the emulator service account key.")
```
Comment (psiemens) [inline internal/emulator/start.go:87]: Does this error message still apply here?

File: internal/emulator/start.go [hunk 2]
```diff
@@ -77,23 +77,22 @@ func ConfiguredServiceKey(
...
 
-	privateKey, err := crypto.DecodePrivateKeyHex(serviceKeyHex.SigAlgo(), serviceKeyHex.PrivateKeyHex())
+	err = serviceAccount.Key().Validate()
 	if err != nil {
-		util.Exitf(
-			1,
-			"Invalid private key in \"%s\" emulator configuration",
-			config.DefaultEmulatorConfigName,
-		)
+		util.Exit(1, err.Error())
```
Comment (psiemens) [inline internal/emulator/start.go:92]: It'd be good to keep the original context of the error message to mention where this error originated from

File: pkg/flowcli/config/json/account.go [hunk 3]
```diff
@@ -39,45 +40,75 @@ func transformAddress(address string) flow.Address {
...
+			SigAlgo:    crypto.ECDSA_P256,
+			HashAlgo:   crypto.SHA3_256,
+			ResourceID: "",
+			PrivateKey: pkey,
+		},
+	}
+}
+
+// transformAdvancedToConfig transforms advanced internal account to config account
+func transformAdvancedToConfig(accountName string, a advanceAccount) config.Account {
```
Comment (psiemens) [inline pkg/flowcli/config/json/account.go:65]:
```suggestion
func transformAdvancedToConfig(accountName string, a advancedAccount) config.Account {
```

File: pkg/flowcli/config/json/account.go [hunk 4]
```diff
@@ -39,45 +40,75 @@ func transformAddress(address string) flow.Address {
...
+			Index:      a.Key.Index,
+			SigAlgo:    sigAlgo,
+			HashAlgo:   crypto.StringToHashAlgorithm(a.Key.HashAlgo),
+			ResourceID: resourceID,
+			PrivateKey: pKey,
+		},
+	}
+}
+
+// todo add error in return and do validation
```
Comment (psiemens) [inline pkg/flowcli/config/json/account.go:101]:
```suggestion
// TODO: Add error in return and do validation
```

I think some editors only pick up todos in this format

File: pkg/flowcli/config/json/account.go [hunk 5]
```diff
@@ -86,107 +117,167 @@ func (j jsonAccounts) transformToConfig() config.Accounts {
...
+	format, err := decideFormat(b)
+	if err != nil {
+		return err
 	}
 
-	// try advanced format
-	var advanced jsonAccountAdvanced
-	err = json.Unmarshal(b, &advanced)
-	if err == nil {
+	// todo refacto switch in array of parsers
```
Comment (psiemens) [inline pkg/flowcli/config/json/account.go:247]:
```suggestion
	// TODO: refactor switch in array of parsers
```

File: pkg/flowcli/project/keys.go [hunk 6]
```diff
@@ -188,20 +180,27 @@ func (a *HexAccountKey) Signer(ctx context.Context) (crypto.Signer, error) {
...
+		SigAlgo:    a.sigAlgo,
+		HashAlgo:   a.hashAlgo,
+		PrivateKey: a.privateKey,
+	}
+}
+
+func (a *HexAccountKey) Validate() error {
+	_, err := crypto.DecodePrivateKeyHex(a.sigAlgo, a.PrivateKeyHex())
+	if err != nil {
+		return fmt.Errorf("invalid private key")
```
Comment (psiemens) [inline pkg/flowcli/project/keys.go:200]:
```suggestion
		return fmt.Errorf("invalid private key: %w", err)
```

**Review by psiemens — APPROVED** [review]
Nice! The code looks much cleaner, too :)

File: internal/emulator/start.go (hunk 1 above)
Comment (sideninja) [reply internal/emulator/start.go:87]: Yes it does, as you potentially could provide an account with the KMS key, and that private key can not be fetched.

**Pull request merged**